rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,192 @@
+data that is sent. By the time the data has reached the code where the
+`component_sent_event_bytes_total` event is emitted, that event has been
+encoded and the actual estimated JSON size has been lost. The sink will need
+to be updated so that when the event is encoded, the encoded event together
+with the pre-encoded JSON bytesize will be sent to the service where the event
+is emitted.
+
+The Kafka sink also currently sends the binary size, but it looks like the
+estimated JSON bytesize is easily accessible at the point of emitting, so would
+not need too much of a change.
+
+To ensure that the correct metric is sent in a type-safe manner, we will wrap
+the estimated JSON size in a newtype:
+
+```rust
+pub struct JsonSize(usize);
+```
+
+The `EventsSent` metric will only accept this type.
+
+### Registered metrics
+
+It is currently not possible to have dynamic tags with preregistered metrics.
+
+We will need to introduce a registered event caching layer that will register
+and cache new events keyed on the tags that are sent to it.
+
+## Rationale
+
+The ability to visualize data flowing through Vector will allow users to ascertain
+the effectiveness of the current use of Vector. This will enable users to
+optimise their configurations to make the best use of Vector's features.
+
+## Drawbacks
+
+The additional tags being added to the metrics will increase the cardinality of
+those metrics if they are enabled.
+
+## Prior Art
+
+
+## Alternatives
+
+We could use an alternative metric instead of estimated JSON size.
+
+- *Network bytes* This provides a more accurate picture of the actual data being received
+  and sent by Vector, but will regularly produce different sizes for an incoming event
+  to an outgoing event.
+- *In memory size* The size of the event as held in memory. This may be more accurate in
+  determining the amount of memory Vector will be utilizing at any time, will often be
+  less accurate compared to the data being sent and received which is often JSON.
+
+## Outstanding Questions
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Add the `source` field to the Event metadata to indicate the source the event has come from.
+- [ ] Update the Volume event metrics to take a `JsonSize` value. Use the compiler to ensure all metrics
+      emitted use this. The `EstimatedJsonEncodedSizeOf` trait will be updated return a `JsonSize`.
+- [ ] Add the Service meaning. Update any sources that potentially create a service to point the meaning
+      to the relevant field.
+- [ ] Introduce an event caching layer that caches registered events based on the tags sent to it.
+- [ ] Update the emitted events to accept the new tags - taking the `telemetry` configuration options
+      into account.
+
+## Future Improvements
+
+- Logs emitted by Vector should also be tagged with `source_id` and `service`.
```

Comment:
Do we know how much effort this would be? If it is on the order of a few days it might be nice to do as part of this work since I can see users expecting it. If it'll take weeks then we can defer it.

Reply:
It's more than we can fit in to the deadline, we agreed to defer for now.